src/backend/executor/execExprInterp.c

```diff
@@ -762,9 +767,17 @@ ExecInterpExpr(ExprState *state, ExprContext *econtext, bool *isnull)
 			}
 			fcinfo->isnull = false;
 			d = op->d.func.fn_addr(fcinfo);
+
 			*op->resvalue = d;
 			*op->resnull = fcinfo->isnull;
 
+			if (sql_dialect == SQL_DIALECT_TSQL && trunc_numeric_result_hook && !(*op->resnull))
+			{
+				func_result_type = get_func_rettype(fcinfo->flinfo->fn_oid);
```

Comment:
just add a hook here and put rest of logic including dialect check inside the hook code, this will reduce code here and no need to include new headers. Applies to other places as well